books/RayTracingTheNextWeek.html

```diff
@@ -2057,105 +2057,147 @@
+    #include <cstdlib>
+    #include <iostream>
 
-            if (!data) {
-                std::cerr << "ERROR: Could not load texture image file '" << filename << "'.\n";
-                width = height = 0;
+    class rtw_image {
+      public:
+        rtw_image() : data(nullptr) {}
+
+        rtw_image(const char* image_filename) {
+            // Loads image data from the given file name. If the RTW_IMAGES environment variable is
+            // defined, looks only in that directory for the image file. If RTW_IMAGES is not
+            // defined, this constructor will search for the specified image file first from the
+            // current directory, then in the images/ subdirectory, then the _parent's_ images/
+            // subdirectory, and then _that_ parent, on so on, for six levels up. If the image was
+            // not loaded successfully, width() and height() will return 0.
+
+            const auto imagedir = getenv("RTW_IMAGES");
+            if (imagedir) {
+                // If RTW_IMAGES is defined, then only use that value to locate the image file.
+                if (load(std::string(imagedir) + "/" + image_filename)) return;
+            } else {
+                // Hunt for the image file in some likely locations.
+                auto filename = std::string(image_filename);
+                if (load(filename)) return;
+                if (load("images/" + filename)) return;
+                if (load("../images/" + filename)) return;
+                if (load("../../images/" + filename)) return;
+                if (load("../../../images/" + filename)) return;
+                if (load("../../../../images/" + filename)) return;
+                if (load("../../../../../images/" + filename)) return;
+                if (load("../../../../../../images/" + filename)) return;
             }
 
-            bytes_per_scanline = bytes_per_pixel * width;
-        }
-
-        ~image_texture() {
-            delete data;
+            std::cerr << "ERROR: Could not load image file '" << image_filename << "'.\n";
+            std::cerr << "STBI: " << stbi_failure_reason() << '\n';
         }
 
-        virtual color value(double u, double v, const vec3& p) const override {
-            // If we have no texture data, then return solid cyan as a debugging aid.
-            if (data == nullptr)
-                return color(0,1,1);
+        ~rtw_image() { STBI_FREE(data); }
 
-            // Clamp input texture coordinates to [0,1] x [1,0]
-            u = interval(0,1).clamp(u);
-            v = 1.0 - interval(0,1).clamp(v);  // Flip V to image coordinates
+        bool load(const std::string filename) {
+            // Loads image data from the given file name. Returns true if the load succeeded.
+            auto components_per_pixel = bytes_per_pixel; // Components per pixel
+            data = stbi_load(filename.c_str(),
+                &image_width, &image_height, &components_per_pixel, bytes_per_pixel);
```

Comment:
consider modifying a local variable instead of member variables,
then
```
if(data -= nullptr)
    return false;

image_width = local_width;
image_height = local_width;
bytes_per_scanline = image_width * bytes_per_pixel;
```

Reply:
The library guarantees these out parameters will be left unchanged on load failure.